src/java.base/share/classes/java/net/URLEncoder.java

```diff
@@ -222,14 +252,133 @@ public static String encode(String s, String enc)
     public static String encode(String s, Charset charset) {
         Objects.requireNonNull(charset, "charset");
 
+        int spaceCount = 0;
+        int needEncodingCount = 0;
+        boolean utf8 = charset == StandardCharsets.UTF_8;
+        int ut8Length = 0;
+        boolean surrogateError = false;
+        for (int i = 0, length = s.length(); i < length; ++i) {
+            char c = s.charAt(i);
+            if (c < 0x80) {
+                if (c == ' ') {
+                    spaceCount++;
+                }
+                if (!dotNeedEncoding(c)) {
+                    needEncodingCount++;
+                }
+            } else {
+                needEncodingCount++;
+                if (utf8) {
+                    if (c < 0x800) {
+                        // 2 bytes
+                        ut8Length += 3;
+                    } else if (Character.isSurrogate(c)) {
+                        // Have a surrogate pair
+                        if (Character.isHighSurrogate(c) && i + 1 < length) {
+                            char d = s.charAt(i + 1);
+                            if (Character.isLowSurrogate(d)) {
+                                ut8Length += 8;
+                                i++;
+                                continue;
+                            }
+                        }
+                        surrogateError = true;
+                        ut8Length += 3;
+                    } else {
+                        // 3 bytes
+                        ut8Length += 6;
+                    }
+                }
+            }
```

Comment:
We can break the loop early if `!utf8 && (needEncodingCount > 0 || spaceCount > 0)`